### PR TITLE
Add IDictionary_Generic_Tests test for multiple values with hash collisions

### PR DESCRIPTION
**Pull request by stephentoub**
Related to https://github.com/dotnet/runtime/issues/33131#issuecomment-716532132
cc: @eiriktsarpalis

### CODE REVIEW

File: src/libraries/System.Collections.Concurrent/tests/ConcurrentDictionary/ConcurrentDictionary.Generic.Tests.cs [hunk 1]
```diff
@@ -85,10 +85,9 @@ public abstract class ConcurrentDictionary_Generic_Tests<TKey, TValue> : IDictio
     {
         #region IDictionary<TKey, TValue Helper Methods
 
-        protected override IDictionary<TKey, TValue> GenericIDictionaryFactory()
-        {
-            return new ConcurrentDictionary<TKey, TValue>();
-        }
+        protected override IDictionary<TKey, TValue> GenericIDictionaryFactory() =>new ConcurrentDictionary<TKey, TValue>();
```
Comment (eiriktsarpalis) [inline src/libraries/System.Collections.Concurrent/tests/ConcurrentDictionary/ConcurrentDictionary.Generic.Tests.cs:88]:
```suggestion
        protected override IDictionary<TKey, TValue> GenericIDictionaryFactory() => new ConcurrentDictionary<TKey, TValue>();
```

File: src/libraries/System.Collections.Concurrent/tests/ConcurrentDictionary/ConcurrentDictionary.Generic.Tests.cs [hunk 2]
```diff
@@ -85,10 +85,9 @@ public abstract class ConcurrentDictionary_Generic_Tests<TKey, TValue> : IDictio
...
+
+        protected override IDictionary<TKey, TValue> GenericIDictionaryFactory(IEqualityComparer<TKey> comparer) => new Dictionary<TKey, TValue>(comparer);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Collections.Concurrent/tests/ConcurrentDictionary/ConcurrentDictionary.Generic.Tests.cs:90]: Unless I'm missing something, shouldn't this be creating a ConcurrentDictionary?

**Review by eiriktsarpalis — APPROVED** [review]
Other than a couple of issues, lgtm. Thanks!

File: src/libraries/System.Collections.Concurrent/tests/ConcurrentDictionary/ConcurrentDictionary.Generic.Tests.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Collections.Concurrent/tests/ConcurrentDictionary/ConcurrentDictionary.Generic.Tests.cs:90]: Ugh, you're not missing anything :-)

File: src/libraries/System.Collections.Concurrent/tests/ConcurrentDictionary/ConcurrentDictionary.Generic.Tests.cs [hunk 3]
```diff
@@ -85,10 +85,9 @@ public abstract class ConcurrentDictionary_Generic_Tests<TKey, TValue> : IDictio
     {
         #region IDictionary<TKey, TValue Helper Methods
 
-        protected override IDictionary<TKey, TValue> GenericIDictionaryFactory()
-        {
-            return new ConcurrentDictionary<TKey, TValue>();
-        }
+        protected override IDictionary<TKey, TValue> GenericIDictionaryFactory() => new ConcurrentDictionary<TKey, TValue>();
+
+        protected override IDictionary<TKey, TValue> GenericIDictionaryFactory(IEqualityComparer<TKey> comparer) => new Dictionary<TKey, TValue>(comparer);
```
Comment (stephentoub) [inline src/libraries/System.Collections.Concurrent/tests/ConcurrentDictionary/ConcurrentDictionary.Generic.Tests.cs:90]:
```suggestion
        protected override IDictionary<TKey, TValue> GenericIDictionaryFactory(IEqualityComparer<TKey> comparer) => new ConcurrentDictionary<TKey, TValue>(comparer);
```

**Pull request merged**